In [5]:
import pandas as pd
import numpy as np
import os


In [6]:
#  creating a string of the path for produt.csv main project folder
path = r"C:/Users/Soni/7-7-23(Instacart Basket Analysis)"

In [7]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)


In [8]:
# import “orders_wrangled.csv” from prepared Data 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'prepared Data', 'orders_wrangled.csv'), index_col = False)


# Data Consistency Checks

### Mixed-Type Data

In [9]:
# create a dataframe
# lets create a new dataframe...

df_test = pd.DataFrame()

In [10]:
# create a mixed data type column
df_test['mix'] = ['a','b','1','True']


In [11]:
# mixed typed data type in column form
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
# Now we have mixed data types ..now check for mixed Types 
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)
        
# The structure that’s being used in this code is called a “for-loop.”
# The “for” in for-loop stands for “for these elements, do this,” and 
# the “loop” describes how the structure works: looping over and over again as it performs the procedures detailed by the “for.”
# for-loop is looping through each column in the dataframe and executing the same block of code each time.

# Within the for-loop, a new variable is created: weird.
# The weird variable will ultimately take a boolean value of either True or False. 
# If True, that means the column contains inconsistent data types.
# If False, that means the column contains only one data type.
# Boolean values can also be represented by numbers: 0 as False and 1 as True.

# Next comes the “if” statement. 
# An if statement checks if some condition is met, and if it’s met, executes a line of code.
# If the condition isn’t met, the code isn’t executed.
# Here, the if statement is checking whether weird is true or false.
# If it’s greater than 0, than it’s true. If not, it’s false.
# If weird is true, the command print(col) is executed, which prints the problematic column for you to see. 
# Because of the for-loop, this command will be executed on every column in your dataframe, printing every mixed-type column it finds.


#  Once you’ve reached a decision, you can run the following code (Mix)

In [13]:
df_test['mix'] = df_test['mix'].astype('str')

In [14]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col,'mixed')
    else:
        print(col,'consistent')

mix consistent


# Missing Values

In [15]:
df_prods.isnull().sum()
# isnull() If you were to use the isnull() function by itself, it would return a value of True or False
# You need to know how many total missing observations there are, which is where the sum() function comes in.
# if returns 1 means True and if 0 thats means False no missing values.
#If every missing observation is equal to 1, then you can simply add them up using the sum() function to obtain the total number of missing observations.


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
#To actually view these 16 values, you can create a subset of the dataframe containing only the values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Addressing Missing Values

In [18]:
#Create a new variable that acts like a flag based on the missing value.
#Impute the value with the mean or median of the column (if the variable is numeric).
#Remove or filter out the missing data.
# if you want to replace with mean then...
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [19]:
# If you choose to use the median then...
df_prods.median()

C:\Users\Soni\AppData\Local\Temp\ipykernel_18204\2074649324.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [20]:
# following code to replace your missing values with mean....
#      df['column with missings'].fillna(mean value, inplace=True)

In [21]:
# following code to replace your missing values with median....
#     df['column with missings'].fillna(median value, inplace=True)

In [22]:
# we cannot replace with mean or median value
# product name is a string value which cannot be replaced with any integer value which we get after calculating mean/median
# You can either remove the missing values entirely
# filter out the ones that aren’t missing into a subset dataframe and continue your analysis with this new dataframe.

## filter out the ones that aren’t missing into a subset dataframe

In [23]:
#The shape function returns the number of rows and columns in a dataframe.
df_prods.shape

(49693, 5)

In [24]:
#create a new dataframe called df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
# df_prods_clean.shape again to check that the number of rows has decreased. 
df_prods_clean.shape

(49677, 5)

In [26]:
# 49,677 is less than 49,693. 
#Your new dataframe should have exactly 16 less rows than the original dataframe
#  Another way to....drop all missing values using this code
#    df_prods.dropna(inplace = True)
#If you wanted to use this command to drop only the NaNs from a particular column
#    df_prods.dropna(subset = [‘product_name’], inplace = True)
#In both cases, rather than creating an entirely new dataframe,
# you’re overwriting df_prods with a new version of df_prods that doesn’t contain the missing values.
#This is done by way of the inplace = True function, which overwrites the original dataframe.
#If you don’t specify an inplace argument in your code, 
#the function will take the default setting, which is inplace = False.
# When specified as False, the command will only return a view of the changed dataframe, 
#leaving the original dataframe untouched.
#overwriting can be risky. Unless you’re absolutely sure it’s safe to drop the values in question

# Duplicates

In [27]:
#Finding Duplicates using this code....

#This code creates a new subset of df_prods_clean—df_dups—containing only rows that are duplicates.
#The duplicated() function is what identifies duplicate rows. It’s run on the df_prods_clean dataframe.
#Any duplicate rows that it finds are saved within the new df_dups dataframe.

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [28]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [29]:
#you’ve identified the duplicate rows within your dataframe, you need to delete them.
# using this -----code-------df.drop_duplicates()

In [30]:
#Before doing so, let’s check the current number of rows in your DataFrame using code---
df_prods_clean.shape

(49677, 5)

In [31]:
#Your df_prods_clean dataframe has 49,677 rows and 5 columns.
#create a new dataframe that doesn’t include the duplicates using this code.....

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

In [33]:
# so you have 49672 rows instead of 49677 so 5 rows has been deleted which were duplicated values.

# 4.5: Data Consistency Checks Question2

**Descriptive Analysis**

*Run the df.describe() function on your df_ords dataframe.*

**df_ords.describe()**

# 2. How to interpret the output of this function describe()?

In [34]:
#df_ords.drop(columns = ['Unnamed'], axis=1)

In [35]:
# Run the df.describe() function on your df_ords dataframe
# Tip: Keep an eye on min and max values!
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [36]:
#df_ords looks normal so we will check with products

In [37]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [38]:
# in df_prods describe() all looks normal--but max value for price looks abnormal--99999.000000 
# 2% Reduced fat Milk cannot have price 99999.0000. which is incorrect. we should correct this mistake.

In [39]:
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices']== 99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

# 3. Check for mixed-type data in your df_ords dataframe

In [40]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col,'mixed')
    else:
        print(col,'consistent')

Unnamed: 0 consistent
order_id consistent
user_id consistent
order_number consistent
orders_day_of_week consistent
order_hour_of_day consistent
days_since_last_order consistent


# 4. If you find mixed-type data, fix it

In [41]:
df_ords.dtypes


Unnamed: 0                 int64
order_id                   int64
user_id                    int64
order_number               int64
orders_day_of_week         int64
order_hour_of_day          int64
days_since_last_order    float64
dtype: object

In [42]:
# All variables have same datatype so we cannot change any datatypes.

# 5. missing values in your df_ords dataframe

In [43]:
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

days_since_prior_order have 206209 missing values but other values have non missing values

In [44]:
#  you can create a subset of the dataframe containing only the values 
df_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [45]:
# view missing value subset
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [46]:
# let see with one user_id=1 deatil information for 'days_since_prior_order' column
df_ords[df_ords['user_id'] == 1]

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   

    order_hour_of_day  days_since_last_order  
0                   8                    NaN  
1                   7                   15.0  
2                  12                   21.0  
3                   7                   29.0  
4                  15                   28.0  
5                   7                   19.0  
6                   9                   20.0  
7                  14                   14.0  
8                  16                    0.0  
9                   8                   30.0  
10                  8                   14.0

**As we see each user_id = 1, order_number =1 has missing value for "days_since_prior_order"**

**Observations: All of the records containing missing values in the 'days_since_prior_order' column also have an 'order_number' value of 1. Because the customer is ordering for the first time, there won't be a value showing how long it has been since a prior order. The maximum value in the 'user_id' column is 206209, which coincides with the number of missing values in the 'days_since_prior_order' column: one for each new customer.**

# 6.Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

**One User_id have one missing value. All together data is too much but if we see per user_id, missing data is just one per id. i donot feel to take any action for these missing values.**

# 7.Run a check for duplicate values in your df_ords data

In [47]:
# check for duplicate values....
df_dups = df_ords[df_ords.duplicated()]

In [48]:
# View Duplicate Values
df_dups 

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [49]:
df_dups.size

0

**No duplicate values found in this data**

# 8.Address the duplicates using an appropriate method.

**No duplicates found so cannot take further actions.**

# 9.Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder

In [52]:
# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [53]:
# Export final df_ords 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))